### PR TITLE
[ML] Add runtime mappings to data frame analytics source config

### LINKED ISSUE #65056: [ML] Add support for search-time runtime fields to data frame analytics
Since #64374 Elasticsearch has the option to have search-time runtime fields.

ML should allow these to be used in data frame analytics.

The most obvious way to do it seems to be to add an optional `runtime_mappings` section to the data frame analytics config with the same structure as the `runtime_mappings` section of searches (as defined in #64374), then add this into the search the data frame analytics job does when reindexing source data to the destination index.

However, after this initial step there is a choice of approaches.  For the destination index we could:

1. Make the search-time runtime fields on the source indices into indexed fields
2. Make the search-time runtime fields on the source indices into runtime fields stored in the destination index mappings
3. Keep the search-time runtime fields on the source indices as search-time runtime fields on the destination index

Some analysis of the pros and cons is needed before deciding which option is best.

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Users can now specify runtime mappings as part of the source config
of a data frame analytics job. Those runtime mappings become part of
the mapping of the destination index. This ensures the fields are
accessible in the destination index even if the relevant data frame
analytics job gets deleted.

Closes #65056

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ml.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public final class RuntimeMappingsValidator {
+
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java:14]: private ctor?
```suggestion

    private RuntimeMappingsValidator() { }

```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public static void validate(Map<String, Object> runtimeMappings) {
+        for (Map.Entry<String, Object> entry : runtimeMappings.entrySet()) {
+            // top level objects are fields
+            String fieldName = entry.getKey();
+            if (entry.getValue() instanceof Map) {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> propNode = new HashMap<>(((Map<String, Object>) entry.getValue()));
+                Object typeNode = propNode.get("type");
+                if (typeNode == null) {
+                    throw ExceptionsHelper.badRequestException("No type specified for runtime field [" + fieldName + "]");
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java:28]:
```suggestion
                    throw ExceptionsHelper.badRequestException("No type specified for runtime field [{}]", fieldName);
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                }
+            } else {
+                throw ExceptionsHelper.badRequestException("Expected map for runtime field [" + fieldName + "] " +
+                    "definition but got a " + fieldName.getClass().getSimpleName());
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/RuntimeMappingsValidator.java:32]:
```suggestion
                throw ExceptionsHelper.badRequestException(
                    "Expected map for runtime field [{}] definition but got a {}", 
                    fieldName, 
                    fieldName.getClass().getSimpleName()
                );
```
I think string formatting like this is much nicer. Feel free to ignore :)

**Review by benwtrent — COMMENTED** [review]
I think it would be good to add a test for regression, classification that utilizes runtime field mappings (didn't see one). This way we can be sure it works there and works fine during the inference phase of analytics.

**Comment by dimitris-athanasiou** [comment]
@benwtrent Thank you for pointing out lack of coverage for regression/classification. Adding tests revealed a couple more places where I had to add runtime mappings. Pushed the fix, should be good for another review now.

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Pull request merged**